WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/AuthorSelectionAdapter.kt

```diff
@@ -0,0 +1,137 @@
+    @Inject lateinit var imageManager: ImageManager
+    @Inject lateinit var uiHelpers: UiHelpers
+
+    private var observers: MutableList<WeakReference<DataSetObserver>> = mutableListOf()
+    private val items = mutableListOf<AuthorFilterListItemUIState>()
+
+    init {
+        (context.applicationContext as WordPress).component().inject(this)
+    }
+
+    override fun getCount(): Int = items.count()
+
+    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
+        var view: View? = convertView
+        val holder: DropdownViewHolder
+
+        if (view == null) {
+            val inflater = LayoutInflater.from(parent.context)
+            view = inflater.inflate(R.layout.post_list_author_selection_dropdown, parent, false)
+            holder = DropdownViewHolder(view)
+            view.tag = holder
+        } else {
+            holder = view.tag as DropdownViewHolder
+        }
+
+        holder.bind(items[position], imageManager, uiHelpers)
+
+        return view!!
+    }
+
+    override fun getItem(position: Int): AuthorFilterListItemUIState = items[position]
+
+    override fun getItemId(position: Int): Long = position.toLong()
+
+    override fun getItemViewType(position: Int): Int = 0
+
+    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+        var view: View? = convertView
+        val holder: NormalViewHolder
+
+        if (view == null) {
+            val inflater = LayoutInflater.from(parent.context)
+            view = inflater.inflate(R.layout.post_list_author_selection, parent, false)
+            holder = NormalViewHolder(view)
+            view.tag = holder
+        } else {
+            holder = view.tag as NormalViewHolder
+        }
+
+        holder.bind(items[position], imageManager, uiHelpers)
+
+        return view!!
+    }
+
+    override fun getViewTypeCount(): Int = 1
+    override fun hasStableIds(): Boolean = false
+    override fun isEmpty(): Boolean = false
+
+    fun notifyDataSetChanged() {
+        for (observer in observers) {
+            observer.get()?.onChanged()
+        }
+    }
+
+    override fun registerDataSetObserver(observer: DataSetObserver) {
+        for (alreadyAdded in observers) {
+            if (observer == alreadyAdded.get()) {
+                // No need to add since it is already added
+                return
+            }
+        }
+
+        observers.add(WeakReference(observer))
+    }
+
+    fun updateItems(newItems: List<AuthorFilterListItemUIState>) {
+        items.clear()
+        items.addAll(newItems)
+        notifyDataSetChanged()
+    }
+
+    override fun unregisterDataSetObserver(observer: DataSetObserver) {
+        for ((index, currentRef) in observers.withIndex()) {
+            if (observer == currentRef.get()) {
+                observers.removeAt(index)
+                return
+            }
+        }
+    }
+
+    private open class NormalViewHolder(protected val itemView: View) {
```

Comment:
Could we use a `sealed class` to better represent these `ViewHolder`s? I believe @malinajirka made a similar suggestion in the previous PR. I have tried it and it looked better to me, but I'd prefer to see your implementation before I share mine.

I'd also suggest improving the names a bit, it took me a while to understand what `Normal` meant.

Reply:
I honestly don't think that using a sealed class adds any benefit here.

Happy to come up with a new name for it though.

Reply:
Sure, it's your call. In that case, just because I mentioned it, here is what I had:

```
// TODO: Change the names
private sealed class ViewHolder(val itemView: View) {
    val image: AppCompatImageView = itemView.findViewById(R.id.post_list_author_selection_image)

    open fun bind(state: AuthorFilterListItemUIState, imageManager: ImageManager, uiHelpers: UiHelpers) {
        val avatarSize = image.resources.getDimensionPixelSize(R.dimen.avatar_sz_small)
        val url = GravatarUtils.fixGravatarUrl(state.avatarUrl, avatarSize)

        imageManager.loadIntoCircle(image, state.imageType, url)
    }

    class NormalViewHolder(itemView: View) : ViewHolder(itemView)

    class DropdownViewHolder(itemView: View) : ViewHolder(itemView) {
        private val text: AppCompatTextView = itemView.findViewById(R.id.post_list_author_selection_text)

        override fun bind(state: AuthorFilterListItemUIState, imageManager: ImageManager, uiHelpers: UiHelpers) {
            super.bind(state, imageManager, uiHelpers)
            val context = itemView.context

            text.text = uiHelpers.getTextOfUiString(context, state.text)

            itemView.setBackgroundColor(ContextCompat.getColor(context, state.dropDownBackground))
        }
    }
}
```

Reply:
Given our [other discussion](https://github.com/wordpress-mobile/WordPress-Android/pull/9484#discussion_r269369488) I'm going to leave this as plain old inheritance.